File: app/src/main/java/io/neoterm/setup/SetupHelper.java
package io.neoterm.setup;

import android.app.Activity;
import android.app.ProgressDialog;
import android.content.Context;
import android.os.Build;

import java.io.File;
import java.util.Arrays;

import io.neoterm.R;
import io.neoterm.frontend.config.NeoTermPath;

/**
 * @author kiva
 */
public final class SetupHelper {
    public static boolean needSetup() {
        final File PREFIX_FILE = new File(NeoTermPath.USR_PATH);
        return !PREFIX_FILE.isDirectory();
    }

    public static void setup(final Activity activity, final SourceConnection connection,
                             final ResultListener resultListener) {
        if (!needSetup()) {
            resultListener.onResult(null);
            return;
        }

        final File prefixFile = new File(NeoTermPath.USR_PATH);

        final ProgressDialog progress = makeProgressDialog(activity);
        progress.setMax(100);
        progress.show();

        new SetupThread(activity, connection, prefixFile, resultListener, progress);
    }

    private static ProgressDialog makeProgressDialog(Context context) {
        return makeProgressDialog(context, context.getString(R.string.installer_message));
    }

    public static ProgressDialog makeProgressDialog(Context context, String message) {
        ProgressDialog dialog = new ProgressDialog(context);
        dialog.setMessage(message);
        dialog.setIndeterminate(false);
        dialog.setCancelable(false);
        dialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL);
        return dialog;
    }

    public static String determineArchName() {
        for (String androidArch : Build.SUPPORTED_ABIS) {
            switch (androidArch) {
                case "arm64-v8a":
                    return "aarch64";
                case "armeabi-v7a":
                    return "arm";
                case "x86_64":
                    return "x86_64";
            }
        }
        throw new RuntimeException("Unable to determine arch from Build.SUPPORTED_ABIS =  " +
                Arrays.toString(Build.SUPPORTED_ABIS));
    }
}

File: app/src/main/java/io/neoterm/setup/SetupHelper.kt
package io.neoterm.setup

import android.app.ProgressDialog
import android.content.Context
import android.os.Build
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import io.neoterm.App
import io.neoterm.R
import io.neoterm.frontend.config.NeoTermPath
import java.io.File
import java.util.*

/**
 * @author kiva
 */
object SetupHelper {
  fun needSetup(): Boolean {
    val PREFIX_FILE = File(NeoTermPath.USR_PATH)
    return !PREFIX_FILE.isDirectory
  }

  fun setup(
    activity: AppCompatActivity, connection: SourceConnection,
    resultListener: ResultListener
  ) {
    if (!needSetup()) {
      resultListener.onResult(null)
      return
    }

    val prefixFile = File(NeoTermPath.USR_PATH)

    val progress = makeProgressDialog(activity)
    progress.max = 100
    progress.show()

    SetupThread(activity, connection, prefixFile, resultListener, progress)
      .start()
  }

  private fun makeProgressDialog(context: Context): ProgressDialog {
    return makeProgressDialog(context, context.getString(R.string.installer_message))
  }

  fun makeProgressDialog(context: Context, message: String): ProgressDialog {
    val dialog = ProgressDialog(context)
    dialog.setMessage(message)
    dialog.isIndeterminate = false
    dialog.setCancelable(false)
    dialog.setProgressStyle(ProgressDialog.STYLE_HORIZONTAL)
    return dialog
  }

  fun makeErrorDialog(context: Context, messageId: Int): AlertDialog {
    return makeErrorDialog(context, context.getString(messageId))
  }

  fun makeErrorDialog(context: Context, message: String): AlertDialog {
    return AlertDialog.Builder(context)
      .setTitle(R.string.error)
      .setMessage(message)
      .setPositiveButton(android.R.string.yes, null)
      .setNeutralButton(R.string.show_help) { _, _ -> App.get().openHelpLink() }
      .create()
  }

  fun determineArchName(): String {
    for (androidArch in Build.SUPPORTED_ABIS) {
      when (androidArch) {
        "arm64-v8a" -> return "aarch64"
        "armeabi-v7a" -> return "arm"
        "x86_64" -> return "x86_64"
      }
    }
    throw RuntimeException(
      "Unable to determine arch from Build.SUPPORTED_ABIS =  "
        + Arrays.toString(Build.SUPPORTED_ABIS)
    )
  }
}
